canon/src/store/mod.rs

```diff
@@ -87,4 +87,14 @@ impl<'a> Source<'a> {
         self.offset += n;
         &self.bytes[old_offset..old_offset + n]
     }
+
+    /// Retrieves the amount of bytes for a varint stored in source
+    pub fn varint_len(&self) -> usize {
+        const MSB: u8 = 0b1000_0000;
+        let mut len = 0;
+        while self.bytes[self.offset + len] & MSB != 0 {
+            len += 1;
+        }
+        len + 1
+    }
```

Comment:
I'm actually not sure this should belong to `Source`: the structure is a generic one used as buffer for any kind of types, where `varint_len` is clearly related only to numbers.

So my suggestion would be move this code inside the `decode` in the macro, since it seems it's the only place where is used.

Also, you could use iterator instead:
```rust
    const MSB: u8 = 0b1000_0000;
    let len = self.bytes.iter().take_while(|b| *b & MSB != 0).count() + 1;
```

Reply:
We need this functionality to actually decide how many bytes to take from the source.

Reply:
It is true that we use this method only in one place and therefore it can make sense to calculate the length of the encoded varint inside of `decode()` within the macro.
However `decode()` doesn't have access to the actual value but only to its encoded version in `source`. So if we calculate the `varint_len` inside of `decode()` we would need to make `source.bytes` and `source.offset` public.

It would look like this:

```rust
fn decode(source: &mut Source) -> Result<Self, $crate::CanonError> {
    const MSB: u8 = 0b000_0000;
    let varint_len = source.bytes[source.offset..].iter().take_while(
        |b| *b & MSB != 0
    ).count() + 1;
    VarInt::decode_var(source.read_bytes(varint_len))
        .map_or(Err(CanonError::InvalidEncoding), |(number, _)| {
            Ok(number)
        })
}
```